app/liquid_tags/liquid_tag_base.rb

```diff
@@ -3,11 +3,57 @@ def self.script
     ""
   end
 
+  def validate_data(parse_context)
+    validate_contexts(parse_context) if context_restricted?
```

Comment:
I'd probably remove the conditionals here and move them into the validation methods ("tell, don't ask"):

```ruby
def validate_contexts(parse_context)
   return unless context_restricted?
   ....
end
```

Since these methods are private you may not even be using them in other places, so it may be easier to just inline them and get rid of the predicate methods:

```ruby
def validate_contexts(parse_context)
  return unless self.class.const_defined?("VALID_CONTEXTS")
  ...
end
```